include/zephyr/kernel/thread.h

```diff
@@ -343,6 +343,11 @@ struct k_thread {
 	struct _pipe_desc pipe_desc;
 #endif
 
+#ifdef CONFIG_SMP
+	/** threads waiting in k_thread_suspend() */
+	_wait_q_t  sync_queue;
```

Comment:
Happy to do the rename.

On unifying "join_queue" and "halt_queue"  ...

Although the retry approach sounds appealing on the surface, it is the complexity surrounding corner cases make me hesitant to go there. If a suspend/resume wakes  threads waiting on k_thread_join(), presumably it is up to k_thread_join() to detect this and perform the retry if desired. This leaves a window where many things may happen. 

1. The timeout could have expired.
2. The target thread could have terminated (before or after the timeout).
3. The k_thread could have been re-used.
 
Getting the behaviour right is tricky.

Alternatively, we could have a smaller field in k_thread that indicates that the waiting thread is waiting to join or suspend and unpend accordingly. This would be cleaner than the retry, but I would expect it to be messier than the separate "halt_queue".